tests/trainer/test_trainer.py

```diff
@@ -1,247 +1,380 @@
-
-
-@pytest.mark.parametrize("hparams_cls", _ALL_LOGGERS_CALLBACKS_ALG_PROFILER_HPARAMS)
-def test_fit_on_all_callbacks_loggers_algs_profilers(
-    composer_trainer_hparams: TrainerHparams,
-    dummy_num_classes: int,
-    hparams_cls: Type[hp.Hparams],
-    monkeypatch: pytest.MonkeyPatch,
-    tmpdir: pathlib.Path,
-):
-    trainer = _build_trainer(composer_trainer_hparams, dummy_num_classes, hparams_cls, monkeypatch, tmpdir)
-    trainer.fit()
-
-
-@pytest.mark.parametrize("hparams_cls", _ALL_LOGGERS_CALLBACKS_ALG_PROFILER_HPARAMS)
-def test_multiple_calls_to_fit(
-    composer_trainer_hparams: TrainerHparams,
-    dummy_num_classes: int,
-    hparams_cls: Type[hp.Hparams],
-    monkeypatch: pytest.MonkeyPatch,
-    tmpdir: pathlib.Path,
-):
-    trainer = _build_trainer(composer_trainer_hparams, dummy_num_classes, hparams_cls, monkeypatch, tmpdir)
-    # idempotency test 1: run the FIT event again
-    trainer.engine.run_event(Event.FIT_START)
-    trainer._train_loop()  # using the private method so we don't shutdown the callbacks
-    trainer.state.max_duration = trainer.state.max_duration + trainer.state.max_duration
-    # idempotency test 2: run FIT again. This will trigger another call to FIT_START
-    if issubclass(hparams_cls, LayerFreezingHparams):
-        pytest.xfail("TODO: Layer freezing does not work with a subsequent call to .fit")
-    trainer.fit()  # using the public method so we do shutdown the callbacks
+from torch.utils.data import DataLoader
+
+from composer.algorithms import LayerFreezing
+from composer.algorithms.cutout.cutout import CutOut
+from composer.callbacks import LRMonitor
+from composer.callbacks.run_directory_uploader import RunDirectoryUploader
+from composer.core.callback import Callback
+from composer.core.types import Model
+from composer.loggers import FileLogger, TQDMLogger, WandBLogger
+from composer.trainer import Trainer
+from composer.trainer.trainer_hparams import algorithms_registry, callback_registry, logger_registry
+from tests.common import (RandomClassificationDataset, RandomImageDataset, SimpleConvModel, SimpleModel, device,
+                          world_size)
+
+
+class TestTrainerInit():
+
+    @pytest.fixture
+    def config(self):
+        return {
+            'model': SimpleModel(),
+            'train_dataloader': DataLoader(dataset=RandomClassificationDataset()),
+            'eval_dataloader': DataLoader(dataset=RandomClassificationDataset()),
+            'max_duration': '2ep',
+        }
+
+    def test_init(self, config):
+        trainer = Trainer(**config)
+        assert isinstance(trainer, Trainer)
+
+    def test_model_ddp_wrapped(self, config, init_process_group):
+        # dist initialized first with init_process_group fixture
+        trainer = Trainer(**config)
+        assert isinstance(trainer.state.model, DistributedDataParallel)
+
+    def test_model_ddp_not_wrapped(self, config):
+        torch.distributed.destroy_process_group()
```

Comment:
Ah, I missed this the first time, but I've run into issues before where destroying the process group was buggy. I don't think it always freed the port, but am not quite sure. I guess if it's reliably working then that is fine.

Reply:
Yeah actually I tried to put in a `autouse=True` to always destroy process groups at the end of a test, but it didn't always free the port, and since the port is encoded as an `envvar` by the composer launch process, it never finds a new free one.